Add support for different insert behaviors

jacoscaz opened the pull request:
## todo

- [X] ArrayStrategy
- [x] BinaryTreeStrategy
- [x] RedBlackTreeStrategy
- [x] Add to README.md

## notes

* Work in progress
* Requires https://github.com/adamhooper/js-sorted-set/pull/6 to be merged first

jacoscaz commented:
@adamhooper I've moved to using the phrasing `insert behavior` instead of `insert strategy` as the latter could get people confused given that the library already uses `strategy` to refer to the underlying indexing mechanism.

adamhooper commented:
Perhaps it would be better to call it `onConflict`? (That makes it sound like the value should be a function; but maybe that's a good idea, with allowed outcomes true/false/throw and the three sensible default functions.)

adamhooper commented:
Ooh ooh, better yet:

`onConflict: (existingValue, newValue) => newValue`

jacoscaz commented:
@adamhooper all right - I've added support for conflict resolution behavior provided as a function, as per your last comment, with the `throw` behavior being the default. I've added tests for all strategies accordingly. Let me know what you think (in due time). I'll update the `README.md` once I get your initial feedback. Have a good weekend!

adamhooper commented on src/SortedSet.coffee:
```diff
@@ -2,16 +2,22 @@ AbstractSortedSet = require('./SortedSet/AbstractSortedSet')
 SortedSet.BinaryTreeStrategy = BinaryTreeStrategy
 SortedSet.RedBlackTreeStrategy = RedBlackTreeStrategy
 
+SortedSet.onInsertConflictThrow = insertConflictResolvers.throw
```
Call it `OnInsertConflictThrow`, for consistency with `ArrayStrategy`, etc.

adamhooper commented on src/SortedSet/ArrayStrategy.coffee:
```diff
@@ -35,15 +36,18 @@ binarySearchForIndex = (array, value, comparator) ->
     index = binarySearchForIndex(@data, value, @comparator)
-    throw 'Value already in set' if @data[index] == value
-    @data.splice(index, 0, value)
+    if @data[index] != undefined and @comparator(@data[index], value) == 0
```
Why change the if-statement condition? (It may be correct; but please write a test that fails without this change.)

jacoscaz replied:
Strict equality comparison doesn't work when, for example, one is working with different objects having identical shapes. I will double-check but I am under the impression that the current tests would already fail without this. I haven't really verified this, though. Thank you for reminding me.

jacoscaz replied:
I've checked - the new tests do indeed fail without changing the `if` condition. The same happens for the `RedBlackTreeStrategy`. I would consider this resolved but the final say is yours.

adamhooper commented on src/SortedSet/BinaryTreeStrategy.coffee:
```diff
@@ -35,19 +36,25 @@ binaryTreeDelete = (node, value, comparator) ->
 
   insert: (value) ->
     compare = @comparator
+    onInsertConflict = @onInsertConflict
```
Nix this line and use `@onInsertConflict` below.

`compare = @comparator` is here to speed up the common code path. This `onInsertConflict = @onInsertConflict` doesn't accomplish that.

adamhooper commented:
Woo-hoo! Thanks for this beautiful contribution.
